jwql/utils/instrument_properties.py

```diff
@@ -0,0 +1,229 @@
+        ymax = np.max(scipix[0]) + 1
+        xmax = np.max(scipix[1]) + 1
+
+        # Adjust the minimum and maximum x and y values if they are within
+        # the reference pixels
+        for key in amp_bounds:
+            bounds = amp_bounds[key]
+            prev_xmin, prev_ymin = bounds[0]
+            prev_xmax, prev_ymax = bounds[1]
+            if prev_xmin < xmin:
+                new_xmin = xmin
+            else:
+                new_xmin = prev_xmin
+            if prev_ymin < ymin:
+                new_ymin = ymin
+            else:
+                new_ymin = prev_ymin
+            if prev_xmax > xmax:
+                new_xmax = xmax
+            else:
+                new_xmax = prev_xmax
+            if prev_ymax > ymax:
+                new_ymax = ymax
+            else:
+                new_ymax = prev_ymax
+            amp_bounds[key] = [(new_xmin, new_ymin), (new_xmax, new_ymax)]
+
+    return num_amps, amp_bounds
+
+
+def calc_frame_time(instrument, aperture, xdim, ydim, amps, sample_time=1.e-5):
+    """Calculate the readout time for a single frame of a given size and
+    number of amplifiers. Note that for NIRISS and FGS, the fast readout
+    direction is opposite to that in NIRCam, so we switch ``xdim`` and
+    ``ydim`` so that we can keep a single equation.
+
+    Parameters:
+    -----------
+    instrument : str
+        Name of the instrument being simulated
+
+    aperture : str
+        Name of aperture being simulated (e.g ``NRCA1_FULL``).
+        Currently this is only used to check for the FGS ``ACQ1``
+        aperture, which uses a unique value of ``colpad`` below.
+
+    xdim : int
+        Number of columns in the frame
+
+    ydim : int
+        Number of rows in the frame
+
+    amps : int
+        Number of amplifiers used to read out the frame
+
+    sample_time : float
+        Time to sample a pixel, in seconds. For NIRCam/NIRISS/FGS
+        this is 10 microseconds = 1e-5 seconds
+
+    Returns:
+    --------
+    frametime : float
+        Readout time in seconds for the frame
+    """
+
+    instrument = instrument.lower()
+    if instrument == "nircam":
+        colpad = 12
+
+        # Fullframe
+        if amps == 4:
+            rowpad = 1
+            fullpad = 1
+        else:
+            # All subarrays
+            rowpad = 2
+            fullpad = 0
+            if ((xdim <= 8) & (ydim <= 8)):
+                # The smallest subarray
+                rowpad = 3
+
+    elif instrument == "niriss":
+        colpad = 12
+
+        # Fullframe
+        if amps == 4:
+            rowpad = 1
+            fullpad = 1
+        else:
+            rowpad = 2
+            fullpad = 0
+
+    elif instrument == 'fgs':
+        colpad = 6
+        if 'acq1' in aperture.lower():
+            colpad = 12
+        rowpad = 1
+        if amps == 4:
+            fullpad = 1
+        else:
+            fullpad = 0
+
+    return ((1.0 * xdim / amps + colpad) * (ydim + rowpad) + fullpad) * sample_time
```

Comment:
From what I can tell, this function only gets run if the aperture is in the short `SUBARRAYS_ONE_OR_FOUR_AMPS` list, but I did notice that if it were ever run for MIRI or NIRSpec it would throw an error. Might be worth it to add a case that throws a `NotImplementedError` or something here for those instruments.

Reply:
I've asked the MIRI and NIRSpec teams for functions to calculate frame times.